lib/record_store/provider/oracle_cloud_dns.rb

```diff
@@ -0,0 +1,179 @@
+require 'oci'
+require 'byebug'
+
+module RecordStore
+  class Provider::OracleCloudDNS < Provider
+    class Error < StandardError; end
+
+    class << self
+      def client
+        @client ||= begin
+          config = OCI::Config.new
+          config.user = secrets['user']
+          config.fingerprint = secrets['fingerprint']
+          config.key_content = secrets['key_content']
+          config.tenancy = secrets['tenancy']
+          config.region = secrets['region']
+          config.validate
+          OCI::Dns::DnsClient.new(config: config)
+        end
+      end
+
+      # Downloads all the records from the provider.
+      #
+      # Returns: an array of `Record` for each record in the provider's zone
+      def retrieve_current_records(zone:, stdout: $stdout)
+        client.get_zone_records(zone)
+          .flat_map { |response| response.data.items }
+          .flat_map { |api_record| build_from_api(api_record) }
+          .compact
+      rescue OCI::Errors::HttpRequestBasedError
+        stdout.puts "Cannot build record for zone: #{zone}"
+        raise
+      end
+
+      # Returns an array of the zones managed by provider as strings
+      def zones
+        client.list_zones(secrets['compartment_id']).data.map(&:name)
+      end
+
+      private
+
+      # Creates a new record to the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def add(record, zone)
+        patch_add_record = [
+          OCI::Dns::Models::RecordOperation.new(
+            domain: record.fqdn,
+            rtype: record.type,
+            ttl: record.ttl,
+            rdata: record.rdata_txt,
+            operation: 'ADD',
+          ),
+        ]
+
+        client.patch_zone_records(
+          zone,
+          OCI::Dns::Models::PatchZoneRecordsDetails.new(items: patch_add_record)
+        )
+      end
+
+      # Deletes an existing record from the zone. It is expected this call modifies external state.
+      #
+      # Arguments:
+      # record - a kind of `Record`
+      def remove(record, zone)
+        record_fqdn = Record.ensure_ends_without_dot(record.fqdn)
+        found_record = client.get_zone_records(
+          zone,
+          rtype: record.type,
+          domain: record_fqdn,
+        ).data.items.select { |r| r.rdata == record.rdata_txt }
+
+        return unless found_record
+        record_hash = found_record.first.record_hash if found_record.length == 1
```

Comment:
`record_hash` is `nil` if `found_record.length != 1` => that might be an issue here... and we might want to raise an error in this case ? Or do we want to call a patch remove operation with a `nil` value for `record_hash` ? What happens in that case ? Does that remove both (what I would expect) or does it just fail (in which case we might want to have a cleaner error ?) ?
Could you maybe add a test with two fully identical record with different hashes ?

And what happens if:
1/ we have added two times the same record (e.g. a `TXT` record with value `SOME_VALUE` and TTL `60` for domain `sometxtrecord.recordstore.io`)
2/ we remove one of the two

What's the state after that ?
This seems very unstable to me, but that might also be because we're not doing a real diff between what exists and what _should_ exist, like the Google Cloud DNS provider does...

Reply:
1/ we cannot add two times the same record as you provided above in OCI.
2/ can't be happened without 1/.

And we are not using `id` for `remove` while we have it for `update` because it's how it works with `record_store`.

Here's an example of the provider template for the `record_store`. (zone needs to be added) Rest of providers also has no id for `delete`.

```
  # Deletes an existing record from the zone. It is expected this call modifies external state.
  #
  # Arguments:
  # record - a kind of `Record`
  def remove(zone, record)
  end

  # Updates an existing record in the zone. It is expected this call modifies external state.
  #
  # Arguments:
  # id - provider specific ID of record to update
  # record - a kind of `Record` which the record with `id` should be updated to
  def update(zone, id, record)
  end
```

Reply:
Added error handling in case `record_hash` is `nil`